core/test/utils/unsort_matrix_test.cpp

```diff
@@ -0,0 +1,163 @@
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/test/utils/unsort_matrix.hpp"
+
+
+#include <cmath>
+#include <random>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/dim.hpp>
+#include <ginkgo/core/base/mtx_io.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/test/utils.hpp"
+
+
+namespace {
+
+
+class UnsortMatrix : public ::testing::Test {
+protected:
+    using value_type = double;
+    using index_type = gko::int32;
+    using Csr = gko::matrix::Csr<value_type, index_type>;
+    using Coo = gko::matrix::Coo<value_type, index_type>;
+    using Dense = gko::matrix::Dense<value_type>;
+    UnsortMatrix()
+        : exec(gko::ReferenceExecutor::create()),
+          rand_engine(42),
+          mtx(gko::initialize<Dense>({{1, 2, 0, 0, 0},
+                                      {0, 0, 0, 0, 0},
+                                      {3, 4, 5, 6, 0},
+                                      {0, 0, 7, 0, 0},
+                                      {0, 0, 8, 9, 10}},
+                                     exec)),
+          empty(Dense::create(exec, gko::dim<2>(0, 0)))
+    {}
+
+    bool is_coo_matrix_sorted(Coo *mtx)
+    {
+        auto size = mtx->get_size();
+        auto vals = mtx->get_values();
+        auto rows = mtx->get_row_idxs();
+        auto cols = mtx->get_col_idxs();
+        auto nnz = mtx->get_num_stored_elements();
+
+        if (nnz <= 0) {
+            return true;
+        }
+
+        auto prev_row = rows[0];
+        auto prev_col = cols[0];
+        for (index_type i = 0; i < nnz; ++i) {
+            auto cur_row = rows[i];
+            auto cur_col = cols[i];
+            if (prev_row == cur_row && prev_col > cur_col) {
+                return false;
+            }
+            prev_row = cur_row;
+            prev_col = cur_col;
+        }
+        return true;
+    }
+
+    std::shared_ptr<const gko::Executor> exec;
+    std::ranlux48 rand_engine;
+    std::unique_ptr<Dense> mtx;
+    std::unique_ptr<Dense> empty;
+};
+
+
+TEST_F(UnsortMatrix, CsrWorks)
+{
+    auto csr = Csr::create(exec);
+    mtx->convert_to(gko::lend(csr));
+    bool was_sorted = csr->is_sorted_by_column_index();
+
+    gko::test::unsort_matrix(gko::lend(csr), rand_engine);
+
+    ASSERT_FALSE(csr->is_sorted_by_column_index());
+    ASSERT_TRUE(was_sorted);
+    GKO_ASSERT_MTX_NEAR(csr, mtx, 0.);
```

Comment:
As the CI output says, since this is the core module you technically cannot use anything which calls a kernel, which I believe multiple parts in what I highlighted do. You should move all of this test to the reference module.

Reply:
Thanks, I implemented the check for a CSR matrix in this test myself to get around that limitation.

Reply:
Is this enough? Don't the conversion (from dense to CSR, here), also call kernels?